component/mimir/rules/rules.go

```diff
@@ -0,0 +1,309 @@
+		Exports: Exports{},
+		Build: func(o component.Options, c component.Arguments) (component.Component, error) {
+			return NewComponent(o, c.(Arguments))
+		},
+	})
+}
+
+type Component struct {
+	log  log.Logger
+	opts component.Options
+	args Arguments
+
+	mimirClient  mimirClient.Interface
+	k8sClient    kubernetes.Interface
+	promClient   promVersioned.Interface
+	ruleLister   promListers.PrometheusRuleLister
+	ruleInformer cache.SharedIndexInformer
+
+	namespaceLister   coreListers.NamespaceLister
+	namespaceInformer cache.SharedIndexInformer
+	informerStopChan  chan struct{}
+	ticker            *time.Ticker
+
+	queue         workqueue.RateLimitingInterface
+	configUpdates chan ConfigUpdate
+
+	namespaceSelector labels.Selector
+	ruleSelector      labels.Selector
+
+	currentState map[string][]mimirClient.RuleGroup
+
+	metrics *metrics
+}
+
+type metrics struct {
+	configUpdatesTotal prometheus.Counter
+
+	eventsTotal   *prometheus.CounterVec
+	eventsFailed  *prometheus.CounterVec
+	eventsRetried *prometheus.CounterVec
+
+	mimirClientTiming *prometheus.HistogramVec
+}
+
+func (m *metrics) Register(r prometheus.Registerer) error {
+	r.MustRegister(
+		m.configUpdatesTotal,
+		m.eventsTotal,
+		m.eventsFailed,
+		m.eventsRetried,
+		m.mimirClientTiming,
+	)
+	return nil
+}
+
+func newMetrics() *metrics {
+	return &metrics{
+		configUpdatesTotal: prometheus.NewCounter(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "config_updates_total",
+			Help:      "Total number of times the configuration has been updated.",
+		}),
+		eventsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_total",
+			Help:      "Total number of events processed, partitioned by event type.",
+		}, []string{"type"}),
+		eventsFailed: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_failed_total",
+			Help:      "Total number of events that failed to be processed, even after retries, partitioned by event type.",
+		}, []string{"type"}),
+		eventsRetried: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "mimir_rules",
+			Name:      "events_retried_total",
+			Help:      "Total number of retries across all events, partitioned by event type.",
+		}, []string{"type"}),
+		mimirClientTiming: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Subsystem: "mimir_rules",
+			Name:      "mimir_client_request_duration_seconds",
+			Help:      "Duration of requests to the Mimir API.",
+			Buckets:   instrument.DefBuckets,
+		}, instrument.HistogramCollectorBuckets),
+	}
+}
+
+type ConfigUpdate struct {
+	args Arguments
+	err  chan error
+}
+
+var _ component.Component = (*Component)(nil)
+var _ component.DebugComponent = (*Component)(nil)
+
+func NewComponent(o component.Options, c Arguments) (*Component, error) {
+	setDefaultArguments(&c)
+
+	metrics := newMetrics()
+	metrics.Register(o.Registerer)
+
+	return &Component{
```

Comment:
Ideally we call `init` here as a way to validate the config file and get any early errors returned to the user before the component is running. However, if `init` will be slow or launches any goroutines, it's best left out and delegated to the Run function.